notebooks/examples/python/BritishNationalGrid.py

```diff
@@ -0,0 +1,370 @@
+
+# COMMAND ----------
+
+count_per_index = uprns_table.groupBy("uprn_bng_500m").count().cache()
+
+# COMMAND ----------
+
+# MAGIC %%mosaic_kepler
+# MAGIC count_per_index "uprn_bng_500m" "bng" 50
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC We will also index our postcodes using a built in generator function.
+
+# COMMAND ----------
+
+postcodes_with_index = (postcodes
+
+                           # We break down the original geometry in multiple smoller mosaic chips, each with its
+                           # own index
+                           .withColumn("mosaic_index", mos.mosaic_explode(col("geometry"), lit(optimal_resolution)))
+
+                           # We don't need the original geometry any more, since we have broken it down into
+                           # Smaller mosaic chips.
+                           .drop("json_geometry", "geometry")
+                          )
+
+# COMMAND ----------
+
+postcodes_with_index.display()
+
+# COMMAND ----------
+
+to_display = postcodes_with_index.select("properties.Name", "mosaic_index.index_id", mos.st_aswkt("mosaic_index.wkb").alias("geometry"))
+
+# COMMAND ----------
+
+# MAGIC %%mosaic_kepler
+# MAGIC to_display "geometry" "geometry(27700)" 200
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ### Performing the spatial join
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC We can now do spatial join between our UPRNs and postcodes.
+
+# COMMAND ----------
+
+with_postcodes = (
+  uprns_table.join(
+    postcodes_with_index,
+    uprns_table["uprn_bng_500m"] == postcodes_with_index["mosaic_index.index_id"],
+    how = "right_outer" # to perserve even emtpy chips
+  ).where(
+    # If the borough is a core chip (the chip is fully contained within the geometry), then we do not need
+    # to perform any intersection, because any point matching the same index will certainly be contained in
+    # the borough. Otherwise we need to perform an st_contains operation on the chip geometry.
+    col("mosaic_index.is_core") | mos.st_contains(col("mosaic_index.wkb"), col("uprn_point"))
+  ).select(
+    "properties.*", "uprn_point", "UPRN", "mosaic_index.index_id", mos.st_aswkt("mosaic_index.wkb").alias("index_geometry")
+  )
+)
+
+display(with_postcodes)
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Visualise the results in Kepler
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC Mosaic abstracts interaction with Kepler in python through mosaic_kepler magic. <br>
+# MAGIC Mosaic_kepler magic takes care of conversion between EPSG:27700 and EPSG:4326 so that Kepler can properly render. <br>
+# MAGIC It can handle columns with bng index ids (int and str formats are both supported) and geometries that are provided in EPSG:27700. Mosaic will convert all the geometries for proper rendering. For CRSIDs that are neither EPSG:27700 nor EPSG:4326 the end user needs to convert the geometries using mos.st_transform(geom, destination_crsid).
```

Comment:
I think that now we accept arbitrary CRSIDs in the "geometry(*)", no need to call st_transform

Reply:
fixed